relayer/chains/substrate/construct.go

```
@@ -0,0 +1,671 @@
+		LatestParaHeight: uint32(paraHead.Number),
+		// TODO: add relaychain to config
+		RelayChain: beefyclienttypes.RelayChain_KUSAMA,
+	}, nil
+}
+
+func beefyAuthorities(blockNumber uint32, conn *rpcclient.SubstrateAPI, method string) ([][]byte, error) {
+	blockHash, err := conn.RPC.Chain.GetBlockHash(uint64(blockNumber))
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch metadata
+	meta, err := conn.RPC.State.GetMetadataLatest()
+	if err != nil {
+		return nil, err
+	}
+
+	storageKey, err := rpcclienttypes.CreateStorageKey(meta, "Beefy", method, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var authorities Authorities
+
+	ok, err := conn.RPC.State.GetStorage(storageKey, &authorities, blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		return nil, fmt.Errorf(ErrBeefyConstructNotFound)
+	}
+
+	// Convert from ecdsa public key to ethereum address
+	var authorityEthereumAddresses [][]byte
+	for _, authority := range authorities {
+		pub, err := crypto.DecompressPubkey(authority[:])
+		if err != nil {
+			return nil, err
+		}
+		ethereumAddress := crypto.PubkeyToAddress(*pub)
+		if err != nil {
+			return nil, err
+		}
+		authorityEthereumAddresses = append(authorityEthereumAddresses, ethereumAddress[:])
+	}
+
+	return authorityEthereumAddresses, nil
+}
+
+func bytes32(bytes []byte) beefyclienttypes.SizedByte32 {
+	var buffer beefyclienttypes.SizedByte32
+	copy(buffer[:], bytes)
+	return buffer
+}
+
+func signedCommitment(
+	conn *rpcclient.SubstrateAPI,
+	blockHash rpcclienttypes.Hash,
+) (rpcclienttypes.SignedCommitment, error) {
+	signedBlock, err := conn.RPC.Chain.GetBlock(blockHash)
+	if err != nil {
+		return rpcclienttypes.SignedCommitment{}, err
+	}
+
+	for _, v := range signedBlock.Justifications {
+		if bytes.Equal(v.ConsensusEngineID[:], []byte("BEEF")) {
+			versionedFinalityProof := &rpcclienttypes.VersionedFinalityProof{}
+
+			err = rpcclienttypes.Decode(v.EncodedJustification, versionedFinalityProof)
```

This error is never checked